drivers/mtd/at25ee.c

```diff
@@ -0,0 +1,1033 @@
+                  geo->neraseblocks = priv->size / priv->blocksize;
+                }
+              else
+                {
+                  geo->blocksize    = priv->pgsize;
+                  geo->erasesize    = priv->pgsize;
+                  geo->neraseblocks = priv->npages;
+                }
+
+              ret               = OK;
+
+              finfo("blocksize: %" PRId32 " erasesize: %" PRId32
+                    " neraseblocks: %" PRId32 "\n",
+                    geo->blocksize, geo->erasesize, geo->neraseblocks);
+            }
+        }
+        break;
+
+      case BIOC_PARTINFO:
+        {
+          FAR struct partition_info_s *info =
+              (FAR struct partition_info_s *)arg;
+          if (info != NULL)
+            {
+              if (priv->blocksize > priv->pgsize)
+                {
+                  info->numsectors  = priv->size / priv->blocksize;
+                  info->sectorsize  = priv->blocksize;
+                }
+              else
+                {
+                  info->numsectors  = priv->npages;
+                  info->sectorsize  = priv->pgsize;
+                }
+
+              info->startsector = 0;
+              info->parent[0]   = '\0';
+              ret               = OK;
+            }
+        }
+        break;
+
+      case MTDIOC_BULKERASE:
+        ret = at25ee_eraseall(priv);
+        break;
+
+      default:
+        ret = -ENOTTY; /* Bad command */
+        break;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: at25ee_initialize
+ *
+ * Description:
+ *   Create an initialized MTD device instance for an AT25 SPI EEPROM
+ *   MTD devices are not registered in the file system, but are created
+ *   as instances that can be bound to other functions
+ *   (such as a block or character driver front end).
+ *
+ * Input Parameters:
+ *   dev        - a reference to the spi device structure
+ *   devtype    - device type, from include/nuttx/eeprom/spi_xx25xx.h
+ *   readonly   - sets block driver to be readonly
+ *
+ * Returned Value:
+ *   Initialised device instance (success) or NULL (fail)
+ *
+ ****************************************************************************/
+
+FAR struct mtd_dev_s *at25ee_initialize(FAR struct spi_dev_s *dev,
+                                        int devtype, int readonly)
+{
+  FAR struct at25ee_dev_s *priv;
+
+  DEBUGASSERT(dev);
+
+  /* Check device type early */
+
+  if ((devtype < 0) ||
+      (devtype >= sizeof(g_at25ee_devices) / sizeof(g_at25ee_devices[0])))
+    {
+      return NULL;
+    }
+
+  priv = kmm_zalloc(sizeof(struct at25ee_dev_s));
+  if (priv == NULL)
+    {
+      ferr("ERROR: Failed to allocate device structure\n");
+      return NULL;
+    }
+
+  if (!priv->initd)
```

Comment:
why check a value which is always 0

Reply:
It isn't always 0? It is set true at the end of the initialization. ALthough I don't think initialization is likely to be called more than once unlike I2C EEPROM so perhaps I just don't need that at all - is that what you're suggesting maybe?

Reply:
I mean initd is always zero at this check point. initd is redundant variable

Reply:
This assumes that at25ee_initialize is only called once , unlike an I2C device which - from the existing code at least - can be unitialised then reinitialised: perhaps as there can be more than one device on the bus, unlike SPI. I will remove initd completely.